zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
+
+        // Queue the block for verification, returning early on error
+        if self.insert_queued_block(block, tx).is_none() {
+            return Box::pin(rx.boxed());
         }
-        .boxed()
-    }
-}
 
-// TODO(teor):
-//   - add a function for the maximum checkpoint height
-//     (We can pre-calculate the result in init(), if we want.)
-//   - check that block.coinbase_height() <= max_checkpoint_height
+        // Try to verify from the next checkpoint to the previous one.
+        //
+        // If this code shows up in profiles, we can try the following optimisations:
+        //   - only check the chain when the length of the HashMap is greater than or equal to
+        //     the length of self.current_checkpoint_range, or
+        //   - cache the height of the last continuous chain as a new field in self, and start
+        //     at that height during the next check.
 
-/// Return a checkpoint verification service, using the provided state service.
-///
-/// The checkpoint verifier holds a state service of type `S`, into which newly
-/// verified blocks will be committed. This state is pluggable to allow for
-/// testing or instrumentation.
-///
-/// The returned type is opaque to allow instrumentation or other wrappers, but
-/// can be boxed for storage. It is also `Clone` to allow sharing of a
-/// verification service.
-///
-/// This function should be called only once for a particular state service (and
-/// the result be shared) rather than constructing multiple verification services
-/// backed by the same state layer.
-pub fn init<S>(
-    state_service: S,
-    checkpoint_list: impl Into<Arc<HashMap<BlockHeight, BlockHeaderHash>>>,
-) -> impl Service<
-    Arc<Block>,
-    Response = BlockHeaderHash,
-    Error = Error,
-    Future = impl Future<Output = Result<BlockHeaderHash, Error>>,
-> + Send
-       + 'static
-where
-    S: Service<zebra_state::Request, Response = zebra_state::Response, Error = Error>
-        + Send
-        + Clone
-        + 'static,
-    S::Future: Send + 'static,
-{
-    CheckpointVerifier {
-        state_service,
-        checkpoint_list: checkpoint_list.into(),
+        // Verification begins with the genesis block, or the first block after
+        // the previous checkpoint.
+        //
+        // The parent of the genesis block has height `None`.
+        let previous_checkpoint_height = self.get_previous_checkpoint_height();
+        // The genesis block's previous hash field is all zeroes
+        // TODO(teor): put this constant somewhere shared between both validators
+        let previous_checkpoint_hash = self
+            .get_previous_checkpoint_hash()
+            .unwrap_or(BlockHeaderHash([0; 32]));
+
+        // If checkpoint verification has finished, we should haved returned when
+        // `insert_queued_block()` failed
+        let (end_height, end_hash) = self.find_checkpoint_chain_end();
+        // TODO(teor): rewrite as a match statement
```

Comment:
fwiw this doesn't look like a great candidate for a match statement

Reply:
An earlier version was much more complicated.